### DISCUSSION TITLE
Offscreen - DisplayHandler.OnCursorChange not work

### DISCUSSION THREAD

**Discussion by zmcx16** [cefsharp/CefSharp#4147] [General]
After I upgraded my Cefsharp.Offscreen from  85.3.130 -> 100.0.140, I found the mouse cursor will be changed on the CefSharp new version when the ChromiumWebBrowser is running, I found this commit:
https://github.com/cefsharp/CefSharp/commit/82581025691ea0f50077a8bb5dc94288d6e0e10b

The MouseCursorChangeDisabled already be removed, and new the OnCursorChange method can be used to control the cursor changed, but I try implment DisplayHandler.OnCursorChange and it doesn't work, the OnCursorChange never be called.

```
    public class CustomDisplayerHandler : DisplayHandler
    {
        protected override bool OnCursorChange(IWebBrowser chromiumWebBrowser, IBrowser browser, IntPtr cursor, CefSharp.Enums.CursorType type, CefSharp.Structs.CursorInfo customCursorInfo)
        {
            // not be called when cursor changed
            Console.WriteLine("cursor changed");
            return false;
        }

        protected override void OnAddressChanged(IWebBrowser chromiumWebBrowser, AddressChangedEventArgs addressChangedArgs)
        {
            // be called when address changed
            Console.WriteLine("address changed");
        }
    }

```

Is there something I missing or anyone know how to disable cursor changed on Cefsharp.Offscreen? Thanks.

**Comment by amaitland** [comment]
> After I upgraded my Cefsharp.Offscreen from 85.3.130 -> 100.0.140

Does the problem reproduce with version 102? The current supported version.

**Reply by amaitland** [reply]
> Is there something I missing or anyone know how to disable cursor changed on Cefsharp.Offscreen?

There's no default cursor implementation in the `OffScreen` version? What do you think your getting by disabling cursor change?

**Reply by zmcx16** [reply]
Hi @amaitland, 

Thank you for replying, version 102.0.100 is the same, sorry for the confusing description, the problem is when I use the Cefsharp.Offscreen to some web crawling, the computer cursor will transfer to "Working In Background" frequently.

https://i.imgur.com/KwuNtCk.png

I would like to make the Cefsharp.Offscreen doesn't change my cursor, but DisplayHandler.OnCursorChange is not working, this problem is not occurring on version 85.3.130. Could you have any suggestions for this problem? Thanks.

**Reply by amaitland** [reply]
I only expect OnCursorChange to be called when the web cursor changes. Are you simulating mouse movements? 

There's no default implementation for OffScreen, so if you are seeing a curse change then it's likely something else causing it.

Possibly comes deep within chromium, or somewhere else in your application. 

The same code path is used in WPF, you can add a breakpoint at the following and use the debugger to validate. When move the mouse over a link the cursor should change. 

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core.Runtime/Internals/ClientAdapter.cpp#L311

Personally I've not seen the cursor change unexpected, at least not that I've noticed, try running the xunit tests in the source and see if you observe the behaviour.

**Reply by zmcx16** [reply]
Hi @amaitland,

Thanks for information, I am not sure this problem is Cefsharp or C# background process behavior, I will keep try how to disable the cursor change to "Working In Background" when the browser load new url.

**Reply by amaitland** [reply]
Did you notice the cursor change when running the xunit tests within the CefSharp source? 
Did you see if cursor change is called when using the WPF version? 

I'd suggest trying to create an isolated example. https://stackoverflow.com/help/minimal-reproducible-example has some guidance. 

From a CefSharp point of view, if there's a mistake in the wrapper code then that's something that can be fixed directly. If the behaviour originates from CEF/chromium then there's nothing I can do. Read first few paragraphs of https://github.com/cefsharp/CefSharp/wiki/General-Usage for background.

**Comment by zmcx16** [comment]
Hi @amaitland,

When I test in NUnit, the cursor doesn't be changed when the browser load new url.
 
I tested on the CefSharp.Winform version:  102.0.100 on a winform application, the cursor still changed to "Working In Background":
https://i.imgur.com/GY23GWy.gif

I also tested on the CefSharp.Winform version:  85.3.130 on a winform application, and I found the cursor also will be changed, but it is not frequently than 102.0.100:
https://i.imgur.com/C5tbMtS.gif

I guess this problem is winform application behavior on a child process rather than the CefSharp, sorry for misleading it...

**Reply by amaitland** [reply]
Do you have UseWaitCursor set somewhere in your application? 

https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.usewaitcursor?view=windowsdesktop-6.0

Is your WinForms UI thread being blocked?

**Reply by zmcx16** [reply]
Hi @amaitland,

The demo program as follow, I think it is not the WinForm UI thread blocking problem, I just set a timer to load url every 2 sec.

https://i.imgur.com/a8BrtLM.gif
```
using CefSharp.WinForms;
using System;
using System.Timers;
using System.Windows.Forms;

namespace CefSharpSample
{
    public partial class Form1 : Form
    {
        public ChromiumWebBrowser _browser;
        public int switch_url_index = 0;

        System.Timers.Timer timer;
        public Form1()
        {
            InitializeComponent();
            FormClosed += this.Form1_FormClosed;
            Cursor.Current = Cursors.WaitCursor;

            _browser = new ChromiumWebBrowser("")
            {
                Dock = DockStyle.Fill,
            };

            SetTimer();

            this.Controls.Add(this._browser);
        }

        private void SetTimer()
        {
            timer = new System.Timers.Timer(2000);
            timer.Elapsed += OnTimedEvent;
            timer.AutoReset = true;
            timer.Enabled = true;
        }

        private void OnTimedEvent(Object source, ElapsedEventArgs e)
        {
            if (switch_url_index % 2 == 0) 
            {
                _browser.Load("https://www.google.com");
            } 
            else 
            {
                _browser.Load("https://www.yahoo.com");
            }
            switch_url_index = (switch_url_index + 1) % 2;
        }

        private void Form1_FormClosed(object sender, FormClosedEventArgs e)
        {
            this._browser?.Dispose();
        }

        private void Form1_Load(object sender, EventArgs e)
        {

        }
    }
}
```

I think this problem probably is CEF rather than CefSharp problem, it is a little annoying but the not a big issue, I would like to let it as a known issue, I appreciate your assistance in helping survey this problem, thanks.

**Reply by amaitland** [reply]
What is the purpose of your demo program? It explicitly sets WaitCursor?

You need to make sure your application has an app.manifest as per https://github.com/cefsharp/CefSharp/blob/cefsharp/102/NuGet/Readme.txt#L13

**Reply by zmcx16** [reply]
Hi @amaitland,

Sorry for misleading, the "Cursor.Current = Cursors.WaitCursor" is just a redundant test code, the cursor will always changed to "Working In Background" when the browser load new url, I also try add the app.manifest but it still not work,

Besides, I found the Console Application the cursor will not be changed to "Working In Background" when the browser load new url. I guess this problem is Winform / WPF problem cursor handle problem, I would like to make it as a known issue, thanks to reply many question.
